### PR TITLE
allow apps to use the service account token for google access

### PR DESCRIPTION
**Pull request by nikosbosse**

### CODE REVIEW

File: crowd-forecast/update-submissions.R [hunk 1]
```diff
@@ -9,9 +9,16 @@ library(here)
 library(tidyr)
 
 # Google sheets authentification -----------------------------------------------
-options(gargle_oauth_cache = ".secrets")
-drive_auth(cache = ".secrets", email = "epiforecasts@gmail.com")
-gs4_auth(token = drive_token())
+# use service account if possible
+path_json <- here::here(".secrets", "crowd-forecast-app-c98ca2164f6c-service-account-token.json")
```
Comment (seabbs) [inline crowd-forecast/update-submissions.R:13]: this is repeated twice so function :)

File: rt-crowd-forecast/extract-samples.R [hunk 2]
```diff
@@ -13,9 +13,16 @@ library(ggplot2)
 library(here)
 
 # Google sheets authentification -----------------------------------------------
-options(gargle_oauth_cache = ".secrets")
-drive_auth(cache = ".secrets", email = "epiforecasts@gmail.com")
-gs4_auth(token = drive_token())
+# use service account if possible
+path_json <- here::here(".secrets", "crowd-forecast-app-c98ca2164f6c-service-account-token.json")
```
Comment (seabbs) [inline rt-crowd-forecast/extract-samples.R:17]: match script style - no `here::`

File: rt-crowd-forecast/update-epinow-data.R [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# update-rt-data by copying it from the Rt folder
+library(here)
```
Comment (seabbs) [inline rt-crowd-forecast/update-epinow-data.R:2]: can we call this update-rt :)

File: rt-crowd-forecast/update-epinow-data.R [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+library(covid.german.forecasts)
+locations <- c("Germany", "Poland")
+
+date <- latest_weekday(Sys.Date())
```
Comment (seabbs) [inline rt-crowd-forecast/update-epinow-data.R:6]: I think this doesn't need Sys.Date()

File: rt-crowd-forecast/update-epinow-data.R [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+for (location in locations) {
+
+  file_names <- c("summarised_estimates.rds", "estimate_samples.rds", 
+                  "model_fit.rds", "model_args.rds", "reported_cases.rds")
+  target_dir <- here("rt-crowd-forecast", "data-raw", "rt-epinow-data", 
+                     location)
+  check_dir(target_dir)
+  for (file_name in file_names) {
+    file.copy(from = here::here("rt-forecast", "data", "samples", "cases",
```
Comment (seabbs) [inline rt-crowd-forecast/update-epinow-data.R:16]: use here:: or just here consistently

**Review by seabbs — CHANGES_REQUESTED** [review]
A genius and a hero.

File: R/utils.R [hunk 6]
```diff
@@ -167,4 +167,36 @@ globalVariables(
...
+#'
+#' @param service_account the path to a JSON file that has all the information
+#' of the Google service account. If a service account is presented it will 
+#' be used and the other arguments will be ignored
+#' @param enail alternatively, provide the email address to an authorised 
+#' account
+#' @param cache_folder path to the folder where secrets for the email address
+#' provided are stored. 
+#' @importFrom here here
+#' @export
```
Comment (seabbs) [inline R/utils.R:182]: missing functions from import.

File: R/utils.R [hunk 7]
```diff
@@ -167,4 +167,36 @@ globalVariables(
...
+  }
+  # if service account is present, use that. Else try authentification via email
+  if (file.exists(service_account)) {
+    gs4_auth(path = service_account)
+  } else {
+    options(gargle_oauth_cache = cache_folder)
+    drive_auth(cache = cache_folder, email = email)
+    gs4_auth(token = drive_token())
+  }
+}
```
Comment (seabbs) [inline R/utils.R:199]: missing return (`return(invisible(NULL))`)

File: R/utils.R [hunk 8]
```diff
@@ -167,4 +167,36 @@ globalVariables(
...
+#' @param enail alternatively, provide the email address to an authorised 
+#' account
+#' @param cache_folder path to the folder where secrets for the email address
+#' provided are stored. 
+#' @importFrom here here
+#' @export
+#' @return NULL
+google_auth <- function(service_account = "default", 
+                        email = "epiforecasts@gmail.com", 
+                        cache_folder = ".secrets") {
```
Comment (seabbs) [inline R/utils.R:186]: service account as arg?

File: crowd-forecast/send-emails.R [hunk 9]
```diff
@@ -16,14 +16,7 @@ if (!file.exists(here::here(".secrets", "epiforecasts-email-creds"))) {
 
 # Google sheets authentification -----------------------------------------------
 # use service account if possible
```
Comment (seabbs) [inline crowd-forecast/send-emails.R:18]: might as well get rid of this comment

File: crowd-forecast/send-emails.R [hunk 10]
```diff
@@ -16,14 +16,7 @@ if (!file.exists(here::here(".secrets", "epiforecasts-email-creds"))) {
...
-path_json <- here::here(".secrets", "crowd-forecast-app-c98ca2164f6c-service-account-token.json")
-if (file.exists(path_json)) {
-  gs4_auth(path = path_json)
-} else {
-  options(gargle_oauth_cache = ".secrets")
-  drive_auth(cache = ".secrets", email = "epiforecasts@gmail.com")
-  gs4_auth(token = drive_token())
-}
+google_auth(service_account = "default")
```
Comment (seabbs) [inline crowd-forecast/send-emails.R:19]: Just call with defaults?

File: R/utils.R [hunk 11]
```diff
@@ -167,4 +167,36 @@ globalVariables(
...
     "type", "value", ".", "primary", "secondary", "target_date", "variable"
   )
-)
\ No newline at end of file
+)
+
+
+
+#' Authentificate for Google Sheets
+#'
```
Comment (seabbs) [inline R/utils.R:173]: does this pass a CRAN check (I don't think we have CI on this repo)

**Review by seabbs — CHANGES_REQUESTED** [review]

File: R/utils.R (hunk 8 above)
Comment (nikosbosse) [reply R/utils.R:189]: I could just have this function without any arguments and have the defaults hard-coded. Happy to change if you think that is better

Comment (seabbs) [reply R/utils.R:189]: ah sorry I didn't realise default led to using the string as listed. This is good.

File: crowd-forecast/app.R [hunk 12]
```diff
@@ -31,6 +31,7 @@ setorder(observations, target_type, target_end_date)
 run_app(data = observations, 
         first_forecast_date = as.character(first_forecast_date),
         selection_vars = c("location_name", "target_type"),
+        path_service_account_json = ".secrets/crowd-forecast-app-c98ca2164f6c-service-account-token.json",
```
Comment (seabbs) [inline crowd-forecast/app.R:34]: have you put this in the `.secrets` folder for each app (i.e in their own folders)? From how I think this is working you need to?

**Review by seabbs — APPROVED** [review]
Looks good to me. Just a q about secret location but otherwise for a merge :)

File: crowd-forecast/app.R (hunk 12 above)
Comment (nikosbosse) [reply crowd-forecast/app.R:34]: yep so currently the secrets are copied three times: once for the overall repo and once for every app

Comment (seabbs) [reply crowd-forecast/app.R:34]: cool that is what I thought.

**Pull request merged**